File: bus_route_optimizer.py
import json
import math
import time
import sys

# Check if running in Google Colab
try:
    import google.colab
    IN_COLAB = True
except ImportError:
    IN_COLAB = False

class BusRouteOptimizer:
    def __init__(self, json_data):
        # The JSON is nested (districts -> sections -> subsections -> routes)
        # We need to extract the flat list of routes first.
        self.routes = self._extract_routes(json_data)
        self.location_index = {}
        self.route_data = []
        self._build_index()

    def _extract_routes(self, data):
        """Recursively extracts all items from lists key'd 'routes'."""
        collected_routes = []
        
        if isinstance(data, dict):
            for key, value in data.items():
                if key == 'routes' and isinstance(value, list):
                    collected_routes.extend(value)
                else:
                    collected_routes.extend(self._extract_routes(value))
        elif isinstance(data, list):
            for item in data:
                collected_routes.extend(self._extract_routes(item))
                
        return collected_routes

    def _build_index(self):
        """Pre-processes the JSON into efficient lookups."""
        print("Building indices...")
        start_time = time.time()
        
        for r_idx, route in enumerate(self.routes):
            # Extract basic info
            route_num = route.get('route_number', 'Unknown')
            
            # Simplified stops: (lat, lon, name, original_obj)
            stops = []
            if 'bus_stops' in route:
                for stop in route['bus_stops']:
                    lat = stop.get('latitude')
                    lon = stop.get('longitude')
                    name = stop.get('location_name', '').lower()
                    if lat is not None and lon is not None:
                        stops.append({
                            'lat': lat,
                            'lon': lon,
                            'name': name,
                            'full_stop': stop
                        })
            
            self.route_data.append({
                'route_number': route_num,
                'stops': stops,
                'original': route
            })

            # Indexing intermediate locations AND bus stop names
            # We normalize to lowercase for case-insensitive search
            keywords = set()
            
            # Add intermediate locations
            if 'intermediate_locations' in route:
                for loc in route['intermediate_locations']:
                    keywords.add(loc.lower())
            
            # Add bus stop names
            for stop in stops:
                keywords.add(stop['name'])
            
            for kw in keywords:
                if kw not in self.location_index:
                    self.location_index[kw] = []
                self.location_index[kw].append(r_idx)
                
        end_time = time.time()
        print(f"Indexing complete in {(end_time - start_time)*1000:.2f} ms")
        print(f"Indexed {len(self.location_index)} unique locations.")

    def haversine_distance(self, lat1, lon1, lat2, lon2):
        """Calculates distance in meters between two points."""
        R = 6371000  # Earth radius in meters
        phi1 = math.radians(lat1)
        phi2 = math.radians(lat2)
        dphi = math.radians(lat2 - lat1)
        dlambda = math.radians(lon2 - lon1)

        a = math.sin(dphi / 2)**2 + \
            math.cos(phi1) * math.cos(phi2) * \
            math.sin(dlambda / 2)**2
        c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))

        return R * c

    def find_best_route(self, user_lat, user_lon, destination_name):
        """
        Finds the best routes to a destination and the closest stop on those routes.
        Returns a sorted list of matches.
        """
        dest_key = destination_name.lower()
        
        # 1. Filter routes that go to the destination
        candidate_indices = self.location_index.get(dest_key, [])
        
        if not candidate_indices:
            return []

        results = []
        
        # 2. For each candidate route, find the closest stop to the user
        for r_idx in candidate_indices:
            route = self.route_data[r_idx]
            stops = route['stops']
            
            if not stops:
                continue
                
            # Find closest stop in this route
            min_dist = float('inf')
            closest_stop = None
            
            for stop in stops:
                dist = self.haversine_distance(user_lat, user_lon, stop['lat'], stop['lon'])
                if dist < min_dist:
                    min_dist = dist
                    closest_stop = stop
            
            if closest_stop:
                results.append({
                    'route_number': route['route_number'],
                    'closest_stop': closest_stop['full_stop'],
                    'distance_meters': min_dist,
                    'destination_match': destination_name
                })
        
        # Sort by distance (closest stop first)
        results.sort(key=lambda x: x['distance_meters'])
        return results

    def find_transfer_route(self, user_lat, user_lon, destination_name):
        """
        Finds a route to the destination. If the pickup point is too far (>1000m),
        looks for a feeder bus to get to that pickup point.
        """
        # 1. Best direct route
        direct_matches = self.find_best_route(user_lat, user_lon, destination_name)
        
        if not direct_matches:
            return None

        best_direct = direct_matches[0]
        dist_to_main_stop = best_direct['distance_meters']
        
        # If close enough, return just this (<= 1000m)
        if dist_to_main_stop <= 1000:
            return {
                "type": "direct",
                "route_1": best_direct
            }
            
        # 2. Too far - try to find a feeder
        # The stop we need to get to is best_direct['closest_stop']
        # We will treat this stop's name as the destination for the feeder bus
        transfer_location = best_direct['closest_stop']['location_name']
        
        # Search for route from User -> Transfer Location
        feeder_matches = self.find_best_route(user_lat, user_lon, transfer_location)
        
        if feeder_matches:
            best_feeder = feeder_matches[0]
            # Verify feeder is actually useful (closer to user than the direct walk)
            # and within reasonable walking distance
            if best_feeder['distance_meters'] < dist_to_main_stop and best_feeder['distance_meters'] < 1000:
                return {
                    "type": "transfer",
                    "route_1": best_feeder,
                    "route_2": best_direct,
                    "transfer_at": transfer_location
                }
                
        # If no better option found, fallback to direct (even if far)
        return {
            "type": "direct_far",
            "route_1": best_direct
        }

# Main Execution Flow
def main():
    json_data = None
    
    if IN_COLAB:
        print("Running in Google Colab.")
        from google.colab import files
        print("Please upload your 'Bus routes.json' file:")
        uploaded = files.upload()
        
        for filename in uploaded.keys():
            print(f"Loading {filename}...")
            # Detect encoding, usually utf-8 but being safe
            try:
                content = uploaded[filename].decode('utf-8')
            except:
                content = uploaded[filename] # already bytes
                
            json_data = json.loads(content)
            break
    else:
        # Local testing path - REPLACE with your actual path if running locally
        file_path = 'Bus routes.json' 
        print(f"Running locally. Looking for {file_path}...")
        try:
            with open(file_path, 'r', encoding='utf-8') as f:
                json_data = json.load(f)
        except FileNotFoundError:
            print(f"Error: {file_path} not found. Please place the file in the directory.")
            return

    if not json_data:
        print("No data loaded.")
        return

    # Initialize Optimizer
    optimizer = BusRouteOptimizer(json_data)

    # --- Benchmarking / Verification ---
    print("\n" + "="*30)
    print("SPEED VERIFICATION")
    print("="*30)
    
    # Test queries
    # Example: User near Pannipitiya (6.85, 79.95), wanting to go to "Malabe"
    user_lat = 6.85
    user_lon = 79.95
    target_dest = "Malabe"
    
    print(f"Scenario: User at ({user_lat}, {user_lon}), Destination: '{target_dest}'")
    
    # Measure single query time with new transfer logic
    start_q = time.time_ns()
    # We now call find_transfer_route instead of find_best_route directly for the final indication
    journey = optimizer.find_transfer_route(user_lat, user_lon, target_dest)
    end_q = time.time_ns()
    
    duration_ms = (end_q - start_q) / 1_000_000
    print(f"Query Time: {duration_ms:.4f} ms")
    
    if duration_ms < 5:
        print("✅ Status: VERY FAST (< 5ms)")
    else:
        print("⚠️ Status: Acceptable")

    if journey:
        j_type = journey['type']
        
        if j_type == 'direct' or j_type == 'direct_far':
            # Standard single bus output
            best = journey['route_1']
            stop = best['closest_stop']
            
            print(f"\n========================================")
            print(f"       RECOMMENDED ROUTE (DIRECT)       ")
            print(f"========================================")
            print(f"Route Number : {best['route_number']}")
            print(f"Destination  : {target_dest}")
            print(f"----------------------------------------")
            print(f"Your Pickup Stop:")
            print(f"  Name       : {stop['location_name']}")
            print(f"  Coordinates: {stop['latitude']}, {stop['longitude']}")
            print(f"  Distance   : {best['distance_meters']:.1f} m")
            print(f"----------------------------------------")
            
            if j_type == 'direct_far':
                print("⚠️ NOTE: The closest stop for this direct route is > 1000m away.")
                print("         We checked for feeder buses but found no better option.")
                
        elif j_type == 'transfer':
            # Two bus output
            feeder = journey['route_1']
            main_route = journey['route_2']
            transfer_loc = journey['transfer_at']
            main_stop = main_route['closest_stop']
            feeder_stop = feeder['closest_stop']
            
            print(f"\n========================================")
            print(f"      RECOMMENDED ROUTE (2 STEPS)       ")
            print(f"========================================")
            print(f"Reason: Direct stop '{transfer_loc}' is {main_route['distance_meters']:.1f}m away (Too far).")
            print(f"========================================")
            
            print(f"STEP 1: FEEDER BUS")
            print(f"  Route      : {feeder['route_number']}")
            print(f"  Action     : Go to pickup stop")
            print(f"  Pickup     : {feeder_stop['location_name']}")
            print(f"  Coordinates: {feeder_stop['latitude']}, {feeder_stop['longitude']}")
            print(f"  Distance   : {feeder['distance_meters']:.1f} m (from your location)")
            print(f"  Drop-off   : {transfer_loc}")
            print(f"----------------------------------------")
            
            print(f"STEP 2: MAIN BUS")
            print(f"  Route      : {main_route['route_number']}")
            print(f"  Action     : Transfer at {transfer_loc}")
            print(f"  Pickup     : {main_stop['location_name']}")
            print(f"  Coordinates: {main_stop['latitude']}, {main_stop['longitude']}")
            print(f"  Destination: {target_dest}")
            print(f"========================================")

    else:
        print("No routes found.")

    # Mass benchmark
    print("\nRunning 10,000 queries to test scalability (using direct route logic)...")
    start_mass = time.time()
    for _ in range(10000):
        optimizer.find_best_route(user_lat, user_lon, target_dest)
    end_mass = time.time()
    avg_time = (end_mass - start_mass) * 1000 / 10000
    print(f"Average time per query: {avg_time:.4f} ms")
    print("="*30)

if __name__ == "__main__":
    main()

File: src/services/BusRouteOptimizer.ts
// File: src/services/BusRouteOptimizer.ts

import * as BusRoutesData from '../data/BusRoutes.json';
import { Bus } from './busService';

// Types defining the structure of the JSON data
interface BusStop {
    combined_code: string;
    global_code: string;
    location_name: string;
    latitude: number;
    longitude: number;
}

interface Route {
    route_number: string;
    start: string;
    end: string;
    intermediate_locations?: string[];
    bus_stops?: BusStop[];
}

interface Node {
    route_number?: string;
    routes?: Route[] | Node[];
    [key: string]: any;
}

// Result types
export interface BusRouteResult {
    routeNumber: string;
    closestStop: BusStop;
    distanceMeters: number;
    destinationMatch: string;
}

export interface JourneyResult {
    type: 'direct' | 'direct_far' | 'transfer';
    route1: BusRouteResult;
    route2?: BusRouteResult;
    transferAt?: string;
    reason?: string;
}

export class BusRouteOptimizer {
    private static instance: BusRouteOptimizer;
    private routes: Route[] = [];
    private locationIndex: Map<string, number[]> = new Map(); // keyword -> route indices
    private routeData: { routeNumber: string; stops: BusStop[]; original: Route }[] = [];
    private initialized = false;

    private constructor() {
        this.init();
    }

    public static getInstance(): BusRouteOptimizer {
        if (!BusRouteOptimizer.instance) {
            BusRouteOptimizer.instance = new BusRouteOptimizer();
        }
        return BusRouteOptimizer.instance;
    }

    private init() {
        if (this.initialized) return;

        // 1. Extract flat list of routes
        this.routes = this.extractRoutes(BusRoutesData as any);

        // 2. Build Index
        this.buildIndex();

        this.initialized = true;
        console.log(`[BusRouteOptimizer] Initialized with ${this.routes.length} routes and ${this.locationIndex.size} indexed locations.`);
    }

    private extractRoutes(data: any): Route[] {
        let collectedRoutes: Route[] = [];

        if (Array.isArray(data)) {
            for (const item of data) {
                collectedRoutes = collectedRoutes.concat(this.extractRoutes(item));
            }
        } else if (typeof data === 'object' && data !== null) {
            for (const key in data) {
                if (key === 'routes' && Array.isArray(data[key])) {
                    collectedRoutes = collectedRoutes.concat(data[key]);
                } else {
                    collectedRoutes = collectedRoutes.concat(this.extractRoutes(data[key]));
                }
            }
        }

        return collectedRoutes;
    }

    private buildIndex() {
        this.routes.forEach((route, index) => {
            // Process stops
            const stops: BusStop[] = [];
            if (route.bus_stops) {
                route.bus_stops.forEach(stop => {
                    if (stop.latitude != null && stop.longitude != null) {
                        stops.push({
                            ...stop,
                            location_name: (stop.location_name || '').toLowerCase()
                        });
                    }
                });
            }

            this.routeData.push({
                routeNumber: route.route_number || 'Unknown',
                stops: stops,
                original: route
            });

            // Index keywords (intermediate locations + stop names)
            const keywords = new Set<string>();

            if (route.intermediate_locations) {
                route.intermediate_locations.forEach(loc => keywords.add(loc.toLowerCase()));
            }

            stops.forEach(stop => keywords.add(stop.location_name));

            keywords.forEach(kw => {
                if (!this.locationIndex.has(kw)) {
                    this.locationIndex.set(kw, []);
                }
                this.locationIndex.get(kw)?.push(index);
            });
        });
    }

    public getIntermediateLocations(): string[] {
        return Array.from(this.locationIndex.keys()).sort();
    }

    public searchLocations(query: string): string[] {
        const q = query.toLowerCase();
        const results: string[] = [];
        // Perform a simple prefix or containment match on keys
        // Since we have direct map access, we iterate keys. 
        // For larger datasets, a trie would be better, but this is fine for <10k locs.
        for (const key of this.locationIndex.keys()) {
            if (key.includes(q)) {
                results.push(key);
                if (results.length > 10) break; // Limit suggestions
            }
        }
        return results;
    }

    private haversineDistance(lat1: number, lon1: number, lat2: number, lon2: number): number {
        const R = 6371000; // meters
        const toRad = (val: number) => val * Math.PI / 180;

        const dLat = toRad(lat2 - lat1);
        const dLon = toRad(lon2 - lon1);

        const a = Math.sin(dLat / 2) * Math.sin(dLat / 2) +
            Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) *
            Math.sin(dLon / 2) * Math.sin(dLon / 2);

        const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        return R * c;
    }

    public findBestRoute(userLat: number, userLon: number, destinationName: string): BusRouteResult[] {
        const destKey = destinationName.toLowerCase();
        const candidateIndices = this.locationIndex.get(destKey);

        if (!candidateIndices) return [];

        const results: BusRouteResult[] = [];

        for (const rIdx of candidateIndices) {
            const routeData = this.routeData[rIdx];
            const stops = routeData.stops;

            if (!stops || stops.length === 0) continue;

            let minDist = Infinity;
            let closestStop: BusStop | null = null;

            for (const stop of stops) {
                const dist = this.haversineDistance(userLat, userLon, stop.latitude, stop.longitude);
                if (dist < minDist) {
                    minDist = dist;
                    closestStop = stop;
                }
            }

            if (closestStop) {
                results.push({
                    routeNumber: routeData.routeNumber,
                    closestStop: closestStop,
                    distanceMeters: minDist,
                    destinationMatch: destinationName
                });
            }
        }

        // Sort by distance
        results.sort((a, b) => a.distanceMeters - b.distanceMeters);
        return results;
    }

    public findTransferRoute(userLat: number, userLon: number, destinationName: string): JourneyResult | null {
        // 1. Best direct route
        const directMatches = this.findBestRoute(userLat, userLon, destinationName);

        if (!directMatches || directMatches.length === 0) {
            return null; // No route found at all
        }

        const bestDirect = directMatches[0];
        const distToMainStop = bestDirect.distanceMeters;

        // Direct is close enough
        if (distToMainStop <= 1000) {
            return {
                type: 'direct',
                route1: bestDirect
            };
        }

        // 2. Too far - try to find a feeder
        // The stop we need to get to
        const transferLocationName = bestDirect.closestStop.location_name;

        // Search for route from User -> Transfer Location
        const feederMatches = this.findBestRoute(userLat, userLon, transferLocationName);

        if (feederMatches && feederMatches.length > 0) {
            const bestFeeder = feederMatches[0];

            // Check if feeder is actually helpful (closer to user than the direct walk)
            // and within reasonable distance
            if (bestFeeder.distanceMeters < distToMainStop && bestFeeder.distanceMeters < 1000) {
                return {
                    type: 'transfer',
                    route1: bestFeeder,
                    route2: bestDirect,
                    transferAt: transferLocationName
                };
            }
        }

        // Fallback if no good feeder found
        return {
            type: 'direct_far',
            route1: bestDirect,
            reason: 'Closest direct stop is > 1000m away, and no suitable feeder bus found.'
        };
    }
    public findRouteToCoordinate(userLat: number, userLon: number, destLat: number, destLon: number): JourneyResult | null {
        // 1. Find the closest known bus stop to the destination coordinates
        let bestStop: BusStop | null = null;
        let minDestDist = Infinity;

        // Iterate through all routes/stops to find the one closest to (destLat, destLon)
        // Optimization: We could spatially index this, but for now linear scan of ~200 stops is fine.
        for (const routeData of this.routeData) {
            if (!routeData.stops) continue;
            for (const stop of routeData.stops) {
                const dist = this.haversineDistance(destLat, destLon, stop.latitude, stop.longitude);
                if (dist < minDestDist) {
                    minDestDist = dist;
                    bestStop = stop;
                }
            }
        }

        if (!bestStop) {
            console.log('No closest stop found for destination coordinates');
            return null;
        }

        // 2. Use that stop's name to find the route from user
        console.log(`Closest stop to saved address is: ${bestStop.location_name} (${Math.round(minDestDist)}m away)`);

        // Use findTransferRoute with the identified stop name
        return this.findTransferRoute(userLat, userLon, bestStop.location_name);
    }
}
